### PR TITLE
Any Descendant

### PR DESCRIPTION
**Pull request by ZwerdDPU**
## Issue
This PR implements descendant wildcard search.

Resolves: #68 #69 #70 #71 #72

## Checklist
- [X] I ran `just verify` locally and it succeeded.
- [X] Issue was given <span style="color: #FF4400">go ahead</span> and is linked above
- [X] Unit tests for my changes are included

### CODE REVIEW

**Comment by V0ldek** [comment]
First draft looks good.

You'll need to add generation of this case to the newly created proptests in query_parser_tests.

**Comment by zwerdlds** [comment]
I've added some proptests for the new selector.

**Comment by zwerdlds** [comment]
I've added some engine as well as determinization tests to validate the any descendant behavior.

File: Justfile [hunk 1]
```diff
@@ -66,6 +66,11 @@ run-debug *ARGS: (build-bin "dev")
 run *ARGS: (build-bin "release")
     ./target/release/rsonpath {{ARGS}}
 
+# === WATCH ===
+watch:
+    cargo watch -x "test -q"
```
[the remark this answers is not in the document]
Comment (zwerdlds) [reply Justfile:71]: This file should not have been added to this commit.  This was super sloppy on my part, apologies.  Unless you disagree, I'll remove this and open a separate issue.

File: crates/rsonpath-lib/src/query/automaton/minimizer.rs [hunk 2]
```diff
@@ -395,6 +395,165 @@ mod tests {
...
+                    fallback_state: State(4),
+                    attributes: StateAttributes::TRANSITIONS_TO_ACCEPTING,
+                },
+                StateTable {
+                    transitions: smallvec![(&label_a, State(5))],
+                    fallback_state: State(4),
+                    attributes: StateAttributes::ACCEPTING,
+                },
+            ],
+        };
```
[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs:487]: > Ahh, your query is incorrect.
> 
> `NfaState::Direct(nfa::Transition::Labelled(&label_b)),`
> 
> This should be a Recursive state!

Should be fixed in the next commit.  Good eye.

File: crates/rsonpath-lib/src/query/automaton/minimizer.rs [hunk 3]
```diff
@@ -395,6 +395,165 @@ mod tests {
...
+                },
+                StateTable {
+                    transitions: smallvec![],
+                    fallback_state: State(2),
+                    attributes: StateAttributes::EMPTY,
+                },
+                StateTable {
+                    transitions: smallvec![],
+                    fallback_state: State(2),
+                    attributes: StateAttributes::EMPTY,
```
[the remark this answers is not in the document]
Comment (V0ldek) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs:424]: > Quite right. Is it possible to add a check on attributes to PartialEq for StateTable?

It's not there? Well then that's a clear bug from me, this should be checked in `PartialEq`. It'd be great if you could add it in this PR.

File: crates/rsonpath-lib/src/query/automaton/minimizer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (V0ldek) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs:487]: I draw them down on a piece of paper/whiteboard and compare. If they look the same up to reassigining the IDs of states then they're equivalent :)

File: crates/rsonpath-lib/src/query/automaton/minimizer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs:424]: Will do.

[the remark this answers is not in the document]
Comment (zwerdlds) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs]: This is resolved in the latest commit.  There were a couple other tests that I had to amend after the change, but the changes made sense.

[the remark this answers is not in the document]
Comment (V0ldek) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs]: Looks good, thanks for fixing this.

File: crates/rsonpath-lib/src/query/automaton/minimizer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (V0ldek) [reply crates/rsonpath-lib/src/query/automaton/minimizer.rs]: Looks good now!

File: Justfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (zwerdlds) [reply Justfile]: Done.

**Comment by zwerdlds** [comment]
I added some more complex tests against the large datasets.

**Review by V0ldek — APPROVED** [review]

**Pull request merged**